File: app/src/main/java/com/healios/dreams/ui/permissions/PermissionViewModel.java
package com.healios.dreams.ui.permissions;

import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;
import androidx.lifecycle.ViewModel;

import com.healios.dreams.data.PermissionRepository;
import com.healios.dreams.model.PermissionModel;

import java.security.Permission;
import java.util.ArrayList;
import java.util.List;

public class PermissionViewModel extends ViewModel {

    private MutableLiveData<List<PermissionModel>> mPermissions;
    private PermissionRepository mPermissionRepository;

    public void init() {
        if (mPermissions != null){
            return;
        }
        mPermissionRepository = PermissionRepository.getInstance();
        mPermissions = mPermissionRepository.getPermissions();
    }

    LiveData<List<PermissionModel>> getPermissions() {
        return mPermissions;
    }

    void updatePermissionStatus(int position, boolean isEnabled){
        List<PermissionModel> permissionModelList = mPermissions.getValue();
        permissionModelList.get(position).setEnabled(isEnabled);
        mPermissions.postValue(permissionModelList);
    }

}

File: app/src/main/java/com/healios/dreams/ui/permissions/PermissionViewModel.kt
package com.healios.dreams.ui.permissions

import android.util.Log
import androidx.lifecycle.LiveData
import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.Observer
import androidx.lifecycle.ViewModel
import com.healios.dreams.data.PermissionRepository
import com.healios.dreams.model.PermissionModel
import com.healios.dreams.util.EmptyLiveEvent
import com.healios.dreams.util.EmptyMutableLiveEvent
import com.healios.dreams.util.Event
import com.healios.dreams.util.managers.PermissionsManager
import com.healios.dreams.util.managers.VersionCompatibilityManager

class PermissionViewModel : ViewModel() {

    private val TAG: String = "PermissionViewModel"
    private var currentPermissionPosition: Int = 0

    private var _permissionList = MutableLiveData<List<PermissionModel>>()
    var permissionList:LiveData<List<PermissionModel>> = _permissionList

    private val mPermissionRepository: PermissionRepository = PermissionRepository.getInstance()
    var canContinue = MutableLiveData(false)

    private val _startEvent = EmptyMutableLiveEvent()
    val startEvent: EmptyLiveEvent = _startEvent
    init {
        _permissionList.value = mPermissionRepository.permissions
        manageGetStartedButtonStatus()
    }

    fun updatePermissionStatus(position: Int, isEnabled: Boolean) {
        val permissionModelList = _permissionList.value!!
        permissionModelList[position].enabled = isEnabled
        _permissionList.value = permissionModelList
        manageGetStartedButtonStatus()
    }

    fun start() {
        _startEvent.value = Event(Unit)
    }

    private fun manageGetStartedButtonStatus() {
        for (permission in _permissionList.value!!) {
            if (!permission.enabled) {
                canContinue.postValue(false)
                return
            }
        }
        canContinue.postValue(true)
    }


}